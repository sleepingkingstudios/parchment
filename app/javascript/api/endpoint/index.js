import defaultGenerateHooks from './hooks';
import defaultGenerateSelector from './selector';
import {
  applyMiddleware,
  selectMiddleware,
} from '../middleware/utils';
import { valueOrDefault } from '../../utils/object';

const applyMiddlewareToRequest = (request, middleware) => {
  request.handleFailure = applyMiddleware(
    request.handleFailure,
    selectMiddleware(middleware, 'handleFailure'),
  );
  request.handlePending = applyMiddleware(
    request.handlePending,
    selectMiddleware(middleware, 'handlePending'),
  );
  request.handleSuccess = applyMiddleware(
    request.handleSuccess,
    selectMiddleware(middleware, 'handleSuccess'),
  );

  return request;
};

class ApiEndpoint {
  constructor(options) {
    const {
      generateActions,
      generateHooks,
      generateInitialState,
      generateReducer,
      generateRequest,
      generateSelector,
      method,
      middleware,
      namespace,
      url,
    } = options;
    const actions = generateActions({ namespace });
    const data = valueOrDefault(options.data, {});
    const initialState = generateInitialState({ data });
    const reducer = generateReducer({ actions, initialState });
    const request = generateRequest({
      actions,
      method,
      namespace,
      url,
    });
    const { performRequest } = request;
    const selector = valueOrDefault(generateSelector, defaultGenerateSelector)({ namespace });
    const hooks = valueOrDefault(generateHooks, defaultGenerateHooks)({
      actions,
      performRequest,
      selector,
    });

    this.actions = actions;
    this.hooks = hooks;
    this.namespace = namespace;
    this.reducer = applyMiddleware(
      reducer,
      selectMiddleware(middleware, 'handleAction'),
    );
    this.request = applyMiddlewareToRequest(request, middleware);
    this.selector = selector;
  }
}

export default ApiEndpoint;
