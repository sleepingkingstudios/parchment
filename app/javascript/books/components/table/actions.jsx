import React from 'react';
import PropTypes from 'prop-types';

import LinkButton from '../../../components/link-button';

const BooksTableActions = ({ id }) => (
  <div className="book-actions">
    <LinkButton
      url={`/books/${id}`}
      link
      buttonSize="sm"
      buttonStyle="info"
    >
      Show
    </LinkButton>
    <LinkButton
      url={`/books/${id}/update`}
      link
      buttonSize="sm"
      buttonStyle="link"
    >
      Update
    </LinkButton>
  </div>
);

BooksTableActions.defaultProps = {};

BooksTableActions.propTypes = {
  id: PropTypes.string.isRequired,
};

export default BooksTableActions;
