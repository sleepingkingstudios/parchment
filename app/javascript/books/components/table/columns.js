import ResponsiveActions from 'components/responsive-actions';
import { injectProps } from 'utils/react';

const generateColumns = ({ actions, baseUrl, useDestroyRequest }) => ([
  {
    label: 'Title',
    prop: 'title',
    width: 5,
  },
  {
    label: 'Publisher',
    prop: 'publisherName',
    width: 4,
  },
  {
    label: false,
    prop: 'actions',
    value: injectProps(
      ResponsiveActions,
      {
        actions,
        baseUrl,
        resourceName: 'book',
        useDestroyRequest,
      },
    ),
    width: 3,
  },
]);

export default generateColumns;
