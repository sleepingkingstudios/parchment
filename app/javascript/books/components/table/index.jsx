/* eslint-disable-next-line import/prefer-default-export */
export { default as columns } from './columns';
