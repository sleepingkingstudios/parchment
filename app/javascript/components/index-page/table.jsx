import React from 'react';
import PropTypes from 'prop-types';
import pluralize from 'pluralize';

import StatusSwitch from '../status-switch';
import Table from '../table';
import { valueOrDefault } from '../../utils/object';
import { underscore } from '../../utils/string';

const defaultMapData = resourceName => data => (data[resourceName]);
const emptyMessage = resourceName => (
  `There are no ${resourceName} matching the criteria.`
);
const renderFailure = (resourceName) => {
  const IndexPageTableFailureMessage = () => (
    <p className="loading-message loading-message-failure">
      Unable to load {resourceName} data from the server.
    </p>
  );

  return IndexPageTableFailureMessage;
};
const renderPending = (resourceName) => {
  const IndexPageTableFailureMessage = () => (
    <p className="loading-message loading-message-failure">
      Loading {resourceName} data from the server...
    </p>
  );

  return IndexPageTableFailureMessage;
};

const IndexPageTable = (props) => {
  const {
    columns,
    endpoint,
    mapData,
    resourceName,
  } = props;
  const pluralResourceName = underscore(pluralize(resourceName));
  const { hooks, request } = endpoint;
  const { useEndpoint } = hooks;
  const { data, status } = useEndpoint();
  const actualMapData = valueOrDefault(
    mapData,
    defaultMapData(pluralResourceName),
  );
  const mappedData = actualMapData(data);
  const { performRequest } = request;
  const onDelete = ({ dispatch, getState }) => {
    performRequest()(dispatch, getState);
  };

  const AppliedIndexPageTable = () => (
    <Table
      columns={columns}
      data={mappedData}
      message={emptyMessage(pluralResourceName)}
      cellProps={{ onDelete }}
    />
  );

  return (
    <StatusSwitch
      renderFailure={renderFailure(pluralResourceName)}
      renderInitialized={renderPending(pluralResourceName)}
      renderPending={renderPending(pluralResourceName)}
      renderSuccess={AppliedIndexPageTable}
      status={status}
    />
  );
};

IndexPageTable.defaultProps = {
  mapData: null,
};

IndexPageTable.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  endpoint: PropTypes.shape({
    hooks: PropTypes.shape({
      useEndpoint: PropTypes.func.isRequired,
    }).isRequired,
    request: PropTypes.shape({
      performRequest: PropTypes.func.isRequired,
    }).isRequired,
  }).isRequired,
  mapData: PropTypes.func,
  resourceName: PropTypes.string.isRequired,
};

export default IndexPageTable;
