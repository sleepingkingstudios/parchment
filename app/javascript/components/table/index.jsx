import React from 'react';
import PropTypes from 'prop-types';

import TableHeader from './header';
import TableMessage from './message';
import TableRow from './row';

const buildBody = (props) => {
  const { columns, data, message } = props;

  if (data.length === 0) {
    return (
      <TableMessage {...{ columns, message }} />
    );
  }

  const { cellProps } = props;

  return data.map(item => (
    <TableRow key={item.id} {...{ columns, item, cellProps }} />
  ));
};

const Table = (props) => {
  const {
    cellProps,
    columns,
    data,
    message,
  } = props;

  return (
    <table className="table">
      <TableHeader columns={columns} />

      <tbody>
        {
          buildBody({
            cellProps,
            columns,
            data,
            message,
          })
        }
      </tbody>
    </table>
  );
};

Table.defaultProps = {
  cellProps: {},
};

Table.propTypes = {
  cellProps: PropTypes.object, /* eslint-disable-line react/forbid-prop-types */
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  message: PropTypes.string.isRequired,
};

export default Table;
