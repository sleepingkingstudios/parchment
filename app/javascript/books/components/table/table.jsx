import IndexPageTable from 'resource/components/table';
import { injectProps } from 'utils/react';

import columns from './columns';

const BooksPageTable = injectProps(
  IndexPageTable,
  { columns },
);

export default BooksPageTable;
