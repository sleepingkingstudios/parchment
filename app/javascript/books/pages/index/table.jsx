import React from 'react';

import StatusSwitch from '../../../components/status-switch';
import { BooksTable } from '../../components/table';
import { hooks, request } from '../../store/indexFindBooks';

const renderFailure = () => (
  <p className="loading-message loading-message-failure">
    Unable to load books data from the server.
  </p>
);
const renderPending = () => (
  <p className="loading-message loading-message-pending">
    Loading books data from the server...
  </p>
);
const { useEndpoint } = hooks;
const { performRequest } = request;
const onDelete = ({ dispatch, getState }) => {
  performRequest()(dispatch, getState);
};

const IndexBooksTable = () => {
  const { data, status } = useEndpoint();
  const { books } = data;

  return (
    <StatusSwitch
      renderFailure={renderFailure}
      renderInitialized={renderPending}
      renderPending={renderPending}
      renderSuccess={() => (<BooksTable books={books} onDelete={onDelete} />)}
      status={status}
    />
  );
};

IndexBooksTable.defaultProps = {};

IndexBooksTable.propTypes = {};

export default IndexBooksTable;
