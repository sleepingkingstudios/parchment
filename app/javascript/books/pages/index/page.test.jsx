import React from 'react';
import { shallow } from 'enzyme';

import IndexBooksPage from './page';
import { columns } from '../../components/table';
import endpoint from '../../store/indexFindBooks';

describe('IndexBooksPage', () => {
  const defaultProps = {};

  it('should render the index page', () => {
    const rendered = shallow(<IndexBooksPage {...defaultProps} />);

    expect(rendered).toHaveDisplayName('IndexPage');
    expect(rendered).toHaveProp({ columns });
    expect(rendered).toHaveProp({ endpoint });
    expect(rendered).toHaveProp({ resourceName: 'Book' });
  });
});
