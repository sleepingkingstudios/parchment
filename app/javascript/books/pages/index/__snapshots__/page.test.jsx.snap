// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksPage should match the snapshot 1`] = `
<Page
  breadcrumbs={
    Array [
      Object {
        "label": "Home",
        "url": "/",
      },
      Object {
        "active": true,
        "label": "Books",
        "url": "/books",
      },
    ]
  }
  className="page-books"
  layout={null}
  subtitle="5e Campaign Companion"
  title="Parchment"
>
  <HeadingWithButtons
    buttons={
      Array [
        Object {
          "label": "Create Book",
          "outline": true,
          "url": "/books/create",
        },
      ]
    }
  >
    Books
  </HeadingWithButtons>
  <IndexBooksTable />
</Page>
`;
